### PR TITLE
[web & desktop] Hide all characters in a TextField, when obscureText is true on web & desktop

### LINKED ISSUE #51218: [web & desktop] Hide all characters in a TextField, when obscureText is true
## Problem
In Flutter Web the last character of a TextField is visible, when obscureText = true. This should not be the case. Specially when your screen is shared with a beamer, etc. everybody can see your password.

Example:
image

## How it should be
Every website hides all characters in a password field.

### PR DESCRIPTION
**Pull request by nilsreichardt**
## Description

Currently a password textfield (obscureText = true) will display the last character. This behavior is expected only on mobile. On web and desktop every character should be hidden (even the last character). With this PR the last character of a password will be only displayed on mobile.

**Demo (Web)**
HideAllCharacters

## Related Issues

Closes #51218

## Tests

* Added a test to check, if the last character is hidden on `TargetPlatform.macOS`, `TargetPlatform.linux`, `TargetPlatform.window`
* Made the existing obscure test specific to `TargetPlatform.android` and `TargetPlatform.iOS`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by AndroidNils** [comment]
@googlebot I fixed it.

**Comment by jonahwilliams** [comment]
> This behavior is expected only on mobile. On web and desktop every character should be hidden

Is this true? I would think that mobile web would show the last character.

**Comment by AndroidNils** [comment]
> > This behavior is expected only on mobile. On web and desktop every character should be hidden
> 
> Is this true? I would think that mobile web would show the last character.

Showing the last character on mobile web would be nice 👍 But Google hides even on mobile web all characters. 

GoogleSignIn

**Comment by jonahwilliams** [comment]
TIL

**Comment by goderbauer** [comment]
In Safari on iOS it does show the last character while you type.

**Comment by justinmc** [comment]
And as mentioned it does show the last character on iOS and Android (not web).  Tagging this as web and Mac.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -2091,10 +2091,16 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
...
       text = widget.obscuringCharacter * text.length;
-      final int o =
-        _obscureShowCharTicksPending > 0 ? _obscureLatestCharIndex : null;
-      if (o != null && o >= 0 && o < text.length)
-        text = text.replaceRange(o, o + 1, _value.text.substring(o, o + 1));
+      // Obscure latest character only on mobile
+      if ((defaultTargetPlatform == TargetPlatform.android ||
+              defaultTargetPlatform == TargetPlatform.iOS ||
+              defaultTargetPlatform == TargetPlatform.fuchsia) &&
+          !kIsWeb) {
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:2098]: It sounds like we should make sure that we get these platforms correct.

@yjbanov You [commented](https://github.com/flutter/flutter/issues/51218#issuecomment-592230646) about this not being an issue on web.  Do you want the behavior to stay as-is on web?  Or @goderbauer mentioned that Safari on iOS shows the last character, do we want to try to match that specific behavior?

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 2]
```diff
@@ -2091,10 +2091,16 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     if (widget.obscureText) {
       String text = _value.text;
       text = widget.obscuringCharacter * text.length;
-      final int o =
-        _obscureShowCharTicksPending > 0 ? _obscureLatestCharIndex : null;
-      if (o != null && o >= 0 && o < text.length)
-        text = text.replaceRange(o, o + 1, _value.text.substring(o, o + 1));
+      // Obscure latest character only on mobile
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/editable_text.dart:2094]: The code inside the if statement is actually revealing the last character, so maybe this is better:

```
// Reveal the latest character in an obscured field only on mobile.
```

That's assuming that it's true that we want to do this only on mobile.  See my comment about platforms.

**Review by justinmc — COMMENTED** [review]
We'll have to figure out the platforms as I commented below.  Otherwise, the main thing this PR needs is a test.  There is already at least one test that verifies the behavior of temporarily showing the last character, see [here](https://github.com/flutter/flutter/blob/7f400af35a0716efc86f5b9401dcbedc501c2ed7/packages/flutter/test/material/text_field_test.dart#L779).

**Comment by goderbauer** [comment]
(PR triage): @AndroidNils Do you have plans to come back to this PR any time soon to address the feedback?

**Comment by AndroidNils** [comment]
@goderbauer Yes, I'll add some tests 👍  Is this point clear for the Flutter team?

> @yjbanov You commented about this not being an issue on web. Do you want the behavior to stay as-is on web? Or @goderbauer mentioned that Safari on iOS shows the last character, do we want to try to match that specific behavior?

(commented by @justinmc)

**Comment by justinmc** [comment]
I think the logic looks correct as-is according to my understanding of what we want for web and desktop 👍

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 2 above)
Comment (AndroidNils) [reply packages/flutter/lib/src/widgets/editable_text.dart]: I have adopted the comment 👍

**Comment by AndroidNils** [comment]
@justinmc 
* Added a test to check, if the last character is hidden on `TargetPlatform.macOS`, `TargetPlatform.linux`, `TargetPlatform.window`
* Made the existing obscure test specific to `TargetPlatform.android` and `TargetPlatform.iOS`

File: packages/flutter/test/material/text_field_test.dart [hunk 3]
```diff
@@ -796,7 +796,46 @@ void main() {
...
+    tester.testTextInput.updateEditingValue(const TextEditingValue(
+      text: testValue + newChar,
+      selection: TextSelection.collapsed(offset: testValue.length + 1),
+    ));
+
+    await tester.pump();
+
+    final String editText = findRenderEditable(tester).text.text;
+    expect(editText.substring(editText.length - 1), '\u2022');
+  }, variant: const TargetPlatformVariant(<TargetPlatform>{
```
Comment (justinmc) [inline packages/flutter/test/material/text_field_test.dart:834]: There are trailing spaces at the end of this line and the following 2 lines, which is causing the [analyzer check to fail](https://github.com/flutter/flutter/pull/56794/checks?check_run_id=739489669).

**Review by justinmc — COMMENTED** [review]
Just needs a quick fix due to trailing spaces, otherwise I think this looks good.  Thanks for making these changes!

**Comment by justinmc** [comment]
Could you also merge master into this?  I think a test is failing because it's out of date.

**Comment by AndroidNils** [comment]
> Could you also merge master into this?  I think a test is failing because it's out of date.

This fixed it 👍

**Review by justinmc — APPROVED** [review]
LGTM. This should be merged when the tree is green.  Thanks again!

**Comment by amichaelrealcme** [comment]
Was this merged into master only and not beta? We're running off of beta for web support, and, last character still shows while typing. Kind of a problem when clients are using zoom screen share and typing in their password. Could this be merged into beta as well?

**Comment by justinmc** [comment]
This went into 1.19.0-5.0.pre and beta is at 1.19.0-4.2.pre as of 2020.06.25.  Hopefully it will get picked up in the next beta release.  Otherwise you could switch to the master channel.

**Pull request merged**